File: frontend/src/lib/api/material.ts
import apiClient from "./client";

// ── Domain enums (mirror backend) ────────────────────────────────────────────

export const MATERIAL_TYPES = [
  "RAW_MATERIAL", "CHEMICAL", "METAL", "PLASTIC", "TEXTILE",
  "ELECTRONIC_COMPONENT", "PACKAGING", "INTERMEDIATE", "COMPOSITE", "OTHER",
] as const;
export type MaterialType = typeof MATERIAL_TYPES[number];

export const MATERIAL_STATUSES = ["ACTIVE", "UNDER_REVIEW", "RESTRICTED", "PHASING_OUT", "ARCHIVED"] as const;
export type MaterialStatus = typeof MATERIAL_STATUSES[number];

export const COMPLIANCE_REGULATIONS = [
  "REACH_SVHC", "ROHS", "CONFLICT_MINERALS", "EUDR", "BATTERY_REGULATION",
  "UFLPA", "CBAM", "POP", "WEEE", "SCIP", "TSCA", "PACKAGING_REGULATION", "CUSTOM",
] as const;
export type ComplianceRegulation = typeof COMPLIANCE_REGULATIONS[number];

export const COMPLIANCE_STATUSES = ["COMPLIANT", "NON_COMPLIANT", "PARTIALLY_COMPLIANT", "UNDER_ASSESSMENT", "EXEMPT", "UNKNOWN"] as const;
export type ComplianceStatus = typeof COMPLIANCE_STATUSES[number];

export const SOURCING_RISKS = ["LOW", "MEDIUM", "HIGH", "CRITICAL"] as const;
export type SourcingRisk = typeof SOURCING_RISKS[number];

// ── Interfaces ────────────────────────────────────────────────────────────────

export interface Material {
  id: string;
  organization_id: string;
  name: string;
  material_type: MaterialType;
  material_status: MaterialStatus;
  internal_code: string | null;
  cas_number: string | null;
  ec_number: string | null;
  iupac_name: string | null;
  molecular_formula: string | null;
  hs_code: string | null;
  un_number: string | null;
  ghs_hazard_class: string | null;
  unit_of_measure: string;
  weight_per_unit_kg: number | null;
  country_of_origin: string | null;
  is_critical_raw_material: boolean;
  recycled_content_pct: number | null;
  description: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface MaterialListResponse {
  items: Material[];
  total: number;
  limit: number;
  offset: number;
}

export interface MaterialComposition {
  id: string;
  organization_id: string;
  parent_material_id: string;
  child_material_id: string;
  weight_pct: number | null;
  quantity: number | null;
  unit: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface MaterialSourcing {
  id: string;
  organization_id: string;
  material_id: string;
  supplier_id: string;
  country_of_origin: string | null;
  annual_volume: number | null;
  unit: string | null;
  price_per_unit_eur: number | null;
  is_primary: boolean;
  lead_time_days: number | null;
  sourcing_risk: SourcingRisk;
  certification_required: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface MaterialCompliance {
  id: string;
  organization_id: string;
  material_id: string;
  regulation: ComplianceRegulation;
  custom_regulation_name: string | null;
  compliance_status: ComplianceStatus;
  assessed_at: string | null;
  valid_until: string | null;
  assessor: string | null;
  evidence_id: string | null;
  notes: string | null;
  is_expired: boolean;
  created_at: string;
  updated_at: string;
}

export interface MaterialSustainability {
  id: string;
  organization_id: string;
  material_id: string;
  reporting_year: number;
  carbon_footprint_kg_co2e_per_kg: number | null;
  carbon_scope: string;
  water_footprint_l_per_kg: number | null;
  energy_mj_per_kg: number | null;
  energy_renewable_pct: number | null;
  recycled_content_pct: number | null;
  recyclability_pct: number | null;
  biodegradable: boolean | null;
  data_source: string | null;
  is_third_party_verified: boolean;
  verification_standard: string | null;
  evidence_id: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

// ── API client functions ──────────────────────────────────────────────────────

// Materials
export const listMaterials = (params?: {
  material_type?: MaterialType;
  material_status?: MaterialStatus;
  search?: string;
  crm_only?: boolean;
  limit?: number;
  offset?: number;
}) => apiClient.get<MaterialListResponse>("/materials", { params });

export const createMaterial = (body: {
  name: string;
  material_type: MaterialType;
  internal_code?: string;
  cas_number?: string;
  ec_number?: string;
  iupac_name?: string;
  molecular_formula?: string;
  hs_code?: string;
  un_number?: string;
  ghs_hazard_class?: string;
  unit_of_measure?: string;
  weight_per_unit_kg?: number;
  country_of_origin?: string;
  is_critical_raw_material?: boolean;
  recycled_content_pct?: number;
  description?: string;
  notes?: string;
}) => apiClient.post<Material>("/materials", body);

export const getMaterial = (id: string) =>
  apiClient.get<Material>(`/materials/${id}`);

export const updateMaterial = (id: string, body: Partial<{
  name: string;
  material_type: MaterialType;
  material_status: MaterialStatus;
  internal_code: string;
  cas_number: string;
  unit_of_measure: string;
  is_critical_raw_material: boolean;
  recycled_content_pct: number;
  description: string;
  notes: string;
}>) => apiClient.put<Material>(`/materials/${id}`, body);

export const archiveMaterial = (id: string) =>
  apiClient.delete(`/materials/${id}`);

// Composition / BOM
export const listComposition = (materialId: string) =>
  apiClient.get<MaterialComposition[]>(`/materials/${materialId}/composition`);

export const addComposition = (materialId: string, body: {
  child_material_id: string;
  weight_pct?: number;
  quantity?: number;
  unit?: string;
  notes?: string;
}) => apiClient.post<MaterialComposition>(`/materials/${materialId}/composition`, body);

export const deleteComposition = (materialId: string, compositionId: string) =>
  apiClient.delete(`/materials/${materialId}/composition/${compositionId}`);

// Sourcing
export const listSourcing = (materialId: string) =>
  apiClient.get<MaterialSourcing[]>(`/materials/${materialId}/sourcing`);

export const addSourcing = (materialId: string, body: {
  supplier_id: string;
  country_of_origin?: string;
  annual_volume?: number;
  unit?: string;
  price_per_unit_eur?: number;
  is_primary?: boolean;
  lead_time_days?: number;
  sourcing_risk?: SourcingRisk;
  certification_required?: string;
  notes?: string;
}) => apiClient.post<MaterialSourcing>(`/materials/${materialId}/sourcing`, body);

export const deleteSourcing = (materialId: string, sourcingId: string) =>
  apiClient.delete(`/materials/${materialId}/sourcing/${sourcingId}`);

// Compliance
export const listCompliance = (materialId: string) =>
  apiClient.get<MaterialCompliance[]>(`/materials/${materialId}/compliance`);

export const upsertCompliance = (materialId: string, body: {
  regulation: ComplianceRegulation;
  compliance_status: ComplianceStatus;
  custom_regulation_name?: string;
  assessed_at?: string;
  valid_until?: string;
  assessor?: string;
  evidence_id?: string;
  notes?: string;
}) => apiClient.post<MaterialCompliance>(`/materials/${materialId}/compliance`, body);

export const deleteComplianceFlag = (materialId: string, flagId: string) =>
  apiClient.delete(`/materials/${materialId}/compliance/${flagId}`);

// Sustainability
export const listSustainability = (materialId: string) =>
  apiClient.get<MaterialSustainability[]>(`/materials/${materialId}/sustainability`);

export const upsertSustainability = (materialId: string, body: {
  reporting_year: number;
  carbon_footprint_kg_co2e_per_kg?: number;
  carbon_scope?: string;
  water_footprint_l_per_kg?: number;
  energy_mj_per_kg?: number;
  energy_renewable_pct?: number;
  recycled_content_pct?: number;
  recyclability_pct?: number;
  biodegradable?: boolean;
  data_source?: string;
  is_third_party_verified?: boolean;
  verification_standard?: string;
  evidence_id?: string;
  notes?: string;
}) => apiClient.post<MaterialSustainability>(`/materials/${materialId}/sustainability`, body);

File: backend/interfaces/api/schemas/material.py
"""API Schemas — Material Twin (M26 / KAN-91–97)."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field

from domain.material import (
    ComplianceRegulation,
    ComplianceStatus,
    MaterialStatus,
    MaterialType,
    SourcingRisk,
)
from .base import EntityResponse


# ── Material Core ─────────────────────────────────────────────────────────────

class MaterialCreate(BaseModel):
    name: str = Field(min_length=1, max_length=500)
    material_type: MaterialType
    internal_code: str | None = Field(default=None, max_length=100)
    cas_number: str | None = Field(default=None, max_length=20)
    ec_number: str | None = Field(default=None, max_length=20)
    iupac_name: str | None = None
    molecular_formula: str | None = Field(default=None, max_length=200)
    hs_code: str | None = Field(default=None, max_length=15)
    un_number: str | None = Field(default=None, max_length=10)
    ghs_hazard_class: str | None = Field(default=None, max_length=200)
    unit_of_measure: str = Field(default="kg", max_length=20)
    weight_per_unit_kg: float | None = Field(default=None, ge=0)
    country_of_origin: str | None = Field(default=None, max_length=100)
    is_critical_raw_material: bool = False
    recycled_content_pct: float | None = Field(default=None, ge=0, le=100)
    description: str | None = None
    notes: str | None = None


class MaterialUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=500)
    material_type: MaterialType | None = None
    material_status: MaterialStatus | None = None
    internal_code: str | None = None
    cas_number: str | None = None
    ec_number: str | None = None
    iupac_name: str | None = None
    hs_code: str | None = None
    unit_of_measure: str | None = Field(default=None, max_length=20)
    weight_per_unit_kg: float | None = Field(default=None, ge=0)
    country_of_origin: str | None = None
    is_critical_raw_material: bool | None = None
    recycled_content_pct: float | None = Field(default=None, ge=0, le=100)
    description: str | None = None
    notes: str | None = None


class MaterialResponse(EntityResponse):
    organization_id: str
    name: str
    material_type: str
    material_status: str
    internal_code: str | None
    cas_number: str | None
    ec_number: str | None
    iupac_name: str | None
    molecular_formula: str | None
    hs_code: str | None
    un_number: str | None
    ghs_hazard_class: str | None
    unit_of_measure: str
    weight_per_unit_kg: float | None
    country_of_origin: str | None
    is_critical_raw_material: bool
    recycled_content_pct: float | None
    description: str | None
    notes: str | None

    @classmethod
    def from_model(cls, m: Any) -> "MaterialResponse":
        return cls(
            id=m.id,
            status=m.status,
            version=m.version,
            created_at=m.created_at,
            updated_at=m.updated_at,
            organization_id=m.organization_id,
            name=m.name,
            material_type=m.material_type,
            material_status=m.material_status,
            internal_code=m.internal_code,
            cas_number=m.cas_number,
            ec_number=m.ec_number,
            iupac_name=m.iupac_name,
            molecular_formula=m.molecular_formula,
            hs_code=m.hs_code,
            un_number=m.un_number,
            ghs_hazard_class=m.ghs_hazard_class,
            unit_of_measure=m.unit_of_measure,
            weight_per_unit_kg=m.weight_per_unit_kg,
            country_of_origin=m.country_of_origin,
            is_critical_raw_material=m.is_critical_raw_material,
            recycled_content_pct=m.recycled_content_pct,
            description=m.description,
            notes=m.notes,
        )


class MaterialListResponse(BaseModel):
    items: list[MaterialResponse]
    total: int
    limit: int
    offset: int


# ── Composition / BOM ─────────────────────────────────────────────────────────

class MaterialCompositionCreate(BaseModel):
    child_material_id: str
    weight_pct: float | None = Field(default=None, ge=0, le=100)
    quantity: float | None = Field(default=None, ge=0)
    unit: str | None = Field(default=None, max_length=20)
    notes: str | None = None


class MaterialCompositionResponse(EntityResponse):
    organization_id: str
    parent_material_id: str
    child_material_id: str
    weight_pct: float | None
    quantity: float | None
    unit: str | None
    notes: str | None

    @classmethod
    def from_model(cls, m: Any) -> "MaterialCompositionResponse":
        return cls(
            id=m.id,
            status=m.status,
            version=m.version,
            created_at=m.created_at,
            updated_at=m.updated_at,
            organization_id=m.organization_id,
            parent_material_id=m.parent_material_id,
            child_material_id=m.child_material_id,
            weight_pct=m.weight_pct,
            quantity=m.quantity,
            unit=m.unit,
            notes=m.notes,
        )


# ── Sourcing ──────────────────────────────────────────────────────────────────

class MaterialSourcingCreate(BaseModel):
    supplier_id: str
    country_of_origin: str | None = Field(default=None, max_length=100)
    annual_volume: float | None = Field(default=None, ge=0)
    unit: str | None = Field(default=None, max_length=20)
    price_per_unit_eur: float | None = Field(default=None, ge=0)
    is_primary: bool = False
    lead_time_days: int | None = Field(default=None, ge=0)
    sourcing_risk: SourcingRisk = SourcingRisk.MEDIUM
    certification_required: str | None = Field(default=None, max_length=100)
    notes: str | None = None


class MaterialSourcingResponse(EntityResponse):
    organization_id: str
    material_id: str
    supplier_id: str
    country_of_origin: str | None
    annual_volume: float | None
    unit: str | None
    price_per_unit_eur: float | None
    is_primary: bool
    lead_time_days: int | None
    sourcing_risk: str
    certification_required: str | None
    notes: str | None

    @classmethod
    def from_model(cls, m: Any) -> "MaterialSourcingResponse":
        return cls(
            id=m.id,
            status=m.status,
            version=m.version,
            created_at=m.created_at,
            updated_at=m.updated_at,
            organization_id=m.organization_id,
            material_id=m.material_id,
            supplier_id=m.supplier_id,
            country_of_origin=m.country_of_origin,
            annual_volume=m.annual_volume,
            unit=m.unit,
            price_per_unit_eur=m.price_per_unit_eur,
            is_primary=m.is_primary,
            lead_time_days=m.lead_time_days,
            sourcing_risk=m.sourcing_risk,
            certification_required=m.certification_required,
            notes=m.notes,
        )


# ── Compliance Flags ──────────────────────────────────────────────────────────

class MaterialComplianceUpsert(BaseModel):
    regulation: ComplianceRegulation
    compliance_status: ComplianceStatus
    custom_regulation_name: str | None = Field(default=None, max_length=200)
    assessed_at: date | None = None
    valid_until: date | None = None
    assessor: str | None = Field(default=None, max_length=300)
    evidence_id: str | None = None
    notes: str | None = None


class MaterialComplianceResponse(EntityResponse):
    organization_id: str
    material_id: str
    regulation: str
    custom_regulation_name: str | None
    compliance_status: str
    assessed_at: date | None
    valid_until: date | None
    assessor: str | None
    evidence_id: str | None
    notes: str | None
    is_expired: bool

    @classmethod
    def from_model(cls, m: Any) -> "MaterialComplianceResponse":
        from datetime import date as _date
        today = _date.today()
        is_expired = m.valid_until is not None and m.valid_until < today
        return cls(
            id=m.id,
            status=m.status,
            version=m.version,
            created_at=m.created_at,
            updated_at=m.updated_at,
            organization_id=m.organization_id,
            material_id=m.material_id,
            regulation=m.regulation,
            custom_regulation_name=m.custom_regulation_name,
            compliance_status=m.compliance_status,
            assessed_at=m.assessed_at,
            valid_until=m.valid_until,
            assessor=m.assessor,
            evidence_id=m.evidence_id,
            notes=m.notes,
            is_expired=is_expired,
        )


# ── Sustainability Metrics ────────────────────────────────────────────────────

class MaterialSustainabilityUpsert(BaseModel):
    reporting_year: int = Field(ge=2000, le=2100)
    carbon_footprint_kg_co2e_per_kg: float | None = Field(default=None, ge=0)
    carbon_scope: str = Field(default="cradle_to_gate", max_length=30)
    water_footprint_l_per_kg: float | None = Field(default=None, ge=0)
    energy_mj_per_kg: float | None = Field(default=None, ge=0)
    energy_renewable_pct: float | None = Field(default=None, ge=0, le=100)
    recycled_content_pct: float | None = Field(default=None, ge=0, le=100)
    recyclability_pct: float | None = Field(default=None, ge=0, le=100)
    biodegradable: bool | None = None
    data_source: str | None = Field(default=None, max_length=300)
    is_third_party_verified: bool = False
    verification_standard: str | None = Field(default=None, max_length=100)
    evidence_id: str | None = None
    notes: str | None = None


class MaterialSustainabilityResponse(EntityResponse):
    organization_id: str
    material_id: str
    reporting_year: int
    carbon_footprint_kg_co2e_per_kg: float | None
    carbon_scope: str
    water_footprint_l_per_kg: float | None
    energy_mj_per_kg: float | None
    energy_renewable_pct: float | None
    recycled_content_pct: float | None
    recyclability_pct: float | None
    biodegradable: bool | None
    data_source: str | None
    is_third_party_verified: bool
    verification_standard: str | None
    evidence_id: str | None
    notes: str | None

    @classmethod
    def from_model(cls, m: Any) -> "MaterialSustainabilityResponse":
        return cls(
            id=m.id,
            status=m.status,
            version=m.version,
            created_at=m.created_at,
            updated_at=m.updated_at,
            organization_id=m.organization_id,
            material_id=m.material_id,
            reporting_year=m.reporting_year,
            carbon_footprint_kg_co2e_per_kg=m.carbon_footprint_kg_co2e_per_kg,
            carbon_scope=m.carbon_scope,
            water_footprint_l_per_kg=m.water_footprint_l_per_kg,
            energy_mj_per_kg=m.energy_mj_per_kg,
            energy_renewable_pct=m.energy_renewable_pct,
            recycled_content_pct=m.recycled_content_pct,
            recyclability_pct=m.recyclability_pct,
            biodegradable=m.biodegradable,
            data_source=m.data_source,
            is_third_party_verified=m.is_third_party_verified,
            verification_standard=m.verification_standard,
            evidence_id=m.evidence_id,
            notes=m.notes,
        )
